src/Symfony/Component/HtmlSanitizer/Visitor/DomVisitor.php

```diff
@@ -0,0 +1,150 @@
+        $this->elementsConfig = $elementsConfig;
+
+        foreach ($config->getAttributesSanitizers() as $attributeSanitizer) {
+            foreach ($attributeSanitizer->getSupportedElements() as $element) {
+                foreach ($attributeSanitizer->getSupportedAttributes() as $attribute) {
+                    $this->attributesSanitizers[$element][$attribute][] = $attributeSanitizer;
+                }
+            }
+        }
+    }
+
+    public function visit(\DOMDocumentFragment $domNode): NodeInterface
+    {
+        $cursor = new Cursor(new DocumentNode());
+        $this->visitChildren($domNode, $cursor);
+
+        return $cursor->node;
+    }
+
+    private function visitNode(\DOMNode $domNode, Cursor $cursor)
+    {
+        $nodeName = u($domNode->nodeName)->lower()->toString();
+
+        // Element should be dropped, including its children
+        if (!\array_key_exists($nodeName, $this->elementsConfig)) {
+            return;
+        }
+
+        // Otherwise, visit recursively
+        $this->enterNode($nodeName, $domNode, $cursor);
+        $this->visitChildren($domNode, $cursor);
+        $cursor->node = $cursor->node->getParent();
+    }
+
+    private function enterNode(string $domNodeName, \DOMNode $domNode, Cursor $cursor)
+    {
+        // Element should be blocked, retaining its children
+        if (false === $this->elementsConfig[$domNodeName]) {
+            $node = new BlockedNode($cursor->node);
+
+            $cursor->node->addChild($node);
+            $cursor->node = $node;
+
+            return;
+        }
+
+        // Otherwise create the node
+        $node = new Node($cursor->node, $domNodeName);
+        $this->setAttributes($domNodeName, $domNode, $node, $this->elementsConfig[$domNodeName]);
+
+        $cursor->node->addChild($node);
+        $cursor->node = $node;
+    }
+
+    private function visitChildren(\DOMNode $domNode, Cursor $cursor)
+    {
+        /** @var \DOMNode $child */
+        foreach ($domNode->childNodes ?? [] as $child) {
+            if ('#text' === $child->nodeName) {
+                // Add text directly for performance
+                $cursor->node->addChild(new TextNode($cursor->node, $child->nodeValue));
+            } elseif (!$child instanceof \DOMText) {
+                // Otherwise continue the visit recursively
+                // Ignore comments for security reasons (interpreted differently by browsers)
+                $this->visitNode($child, $cursor);
+            }
+        }
+    }
+
+    /**
+     * Set attributes from a DOM node to a sanitized node.
+     */
+    private function setAttributes(string $domNodeName, \DOMNode $domNode, Node $node, array $allowedAttributes = [])
+    {
+        if (!$domNode->attributes || !$domNode->attributes->count()) {
+            return;
+        }
+
+        /** @var \DOMAttr $attribute */
+        foreach ($domNode->attributes as $attribute) {
+            $name = u($attribute->name)->lower()->toString();
```

Comment:
from the rfc, lowercasing should happen only on A-Z, not on other unicode chars:
https://w3c.github.io/html-reference/terminology.html#case-insensitive